pandas/core/algorithms.py

```diff
@@ -480,9 +480,32 @@ def _factorize_array(
     return codes, uniques
 
 
-_shared_docs[
-    "factorize"
-] = """
+@doc(
+    values=dedent(
```

Comment:
Would it be possible to get rid of the need for those `dedent` calls ? (thinking about how to make this even easier)

I suppose one option is *writing* those dedented. Something like 

``` 
@doc(
    values=(
        """\
values : sequence
    A 1-D sequence. Sequences that aren't pandas objects are
    coerced to ndarrays before factorization.
        """), 
...
)
def factorize(...
```

but we might consider this ugly? (and does Black allow it?)

And in another this dedent could be handled by the decorator?

Reply:
+1 on adding indentation "magic" to the decorator. Not a big fan of automatic things, but I think making the docstring injection as simple as possible, so the important code is more readable, makes it worth.

Reply:
I think it is a good idea to add this "magic" to decorator so we can keep doc simple.

I am considering change from
```python
wrapper.__doc__ = wrapper._docstr_template.format(**kwargs)  # type: ignore
```
to add `dedent`to all kwargs like this
```python
wrapper.__doc__ = wrapper._docstr_template.format(**{k: dedent(v) for k, v in kwargs.items()})  # type: ignore
```

Any thoughts? Comments? Please.

Reply:
That sounds good. I am only wondering if there are some cases where the to-be-inserted string starts with a space that we want to keep (to let it fit in the docstring)

Reply:
There are some cases where we don't do a `dedent` call, like in https://github.com/pandas-dev/pandas/blob/master/pandas/core/groupby/groupby.py#L66. So you might need to ensure something automatic works there as well.

Reply:
> That sounds good. I am only wondering if there are some cases where the to-be-inserted string starts with a space that we want to keep (to let it fit in the docstring)

Good point. The changes we made here should not change the actual docstring. Since there are mixed cases of usage, it seems hard for me to get a good solution for all of them. At least nothing comes out in my mind now.

I like the "magic", but for now, maybe it is better to keep it without a `dedent` call. Then, we can manually control what shows there. Also, if we do not add `dedent` we can change the use of `Appender` and `Substitution` to `doc` easily and straightforward, because the `doc`'s behave is very close to simply combine `Appender` and `Substitution` together. 

How about we just keep this unchanged now, and we can come back later when we have converted most decorator usages to `doc`? At that time, we might have more information to decided what could be the best way of solving that.

Reply:
Sounds good, I also think it's better to start simple, see how things work, and add more features later.

I think adding the automatic indentation will be worth, since it'll make things much simpler when documenting, and it's not so complex to implement in the decorator. But let's forget about it for now, this is already a nice improvement.

Thanks!